egs/wsj/s5/steps/nnet3/nnet3_train_lib.py

```diff
@@ -0,0 +1,618 @@
+        if (feat_dim != egs_feat_dim) or (ivector_dim != egs_ivector_dim):
+            raise Exception('There is mismatch between featdim/ivector_dim of the current experiment and the provided egs directory')
+
+        if (egs_left_context < left_context) or (egs_right_context < right_context):
+            raise Exception('The egs have insufficient context')
+
+        frames_per_eg = int(open('{0}/info/frames_per_eg'.format(egs_dir)).readline())
+        num_archives = int(open('{0}/info/num_archives'.format(egs_dir)).readline())
+
+        return [egs_left_context, egs_right_context, frames_per_eg, num_archives]
+    except IOError, ValueError:
+        raise Exception('The egs dir {0} has missing or malformed files'.format(egs_dir))
+
+def ComputePreconditioningMatrix(dir, egs_dir, num_lda_jobs, run_opts,
+                                 max_lda_jobs = None, rand_prune = 4.0,
+                                 lda_opts = None):
+    if max_lda_jobs is not None:
+        if num_lda_jobs > max_lda_jobs:
+            num_lda_jobs = max_lda_jobs
+
+    RunKaldiCommand("""
+{command} JOB=1:{num_lda_jobs} {dir}/log/get_lda_stats.JOB.log \
+ nnet3-acc-lda-stats --rand-prune={rand_prune} \
+    {dir}/init.raw "ark:{egs_dir}/egs.JOB.ark" {dir}/JOB.lda_stats""".format(
+        command = run_opts.command,
+        num_lda_jobs = num_lda_jobs,
+        dir = dir,
+        egs_dir = egs_dir,
+        rand_prune = rand_prune))
+
+    # the above command would have generated dir/{1..num_lda_jobs}.lda_stats
+    lda_stat_files = map(lambda x: '{0}/{1}.lda_stats'.format(dir, x),
+                         range(1, num_lda_jobs + 1))
+   
+    RunKaldiCommand("""
+{command} {dir}/log/sum_transform_stats.log \
+    sum-lda-accs {dir}/lda_stats {lda_stat_files}""".format(
+        command = run_opts.command,
+        dir = dir, lda_stat_files = " ".join(lda_stat_files)))
+
+    for file in lda_stat_files:
+        try:
+            os.remove(file)
+        except OSError:
+            raise Exception("There was error while trying to remove lda stat files.")
+    # this computes a fixed affine transform computed in the way we described in
+    # Appendix C.6 of http://arxiv.org/pdf/1410.7455v6.pdf; it's a scaled variant
+    # of an LDA transform but without dimensionality reduction.
+
+    RunKaldiCommand("""
+{command} {dir}/log/get_transform.log \
+ nnet-get-feature-transform {lda_opts} {dir}/lda.mat {dir}/lda_stats
+     """.format(command = run_opts.command,dir = dir,
+                lda_opts = lda_opts if lda_opts is not None else ""))
+            
+    ForceSymlink("../lda.mat", "{0}/configs/lda.mat".format(dir))
+
+import os, errno
+
+def ForceSymlink(file1, file2):
+    try:
+        os.symlink(file1, file2)
+    except OSError, e:
+        if e.errno == errno.EEXIST:
+            os.remove(file2)
+            os.symlink(file1, file2)
+
+def ComputePresoftmaxPriorScale(dir, alidir, num_jobs, run_opts,
```

Comment:
We are using it in the TDNN recipes. This file is shared by both train_dnn.py and train_rnn.py